agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp

```diff
@@ -897,13 +897,36 @@ bool backhaul_manager::backhaul_fsm_main(bool &skip_select)
             if (db->backhaul.connection_type == AgentDB::sBackhaul::eConnectionType::Wireless) {
                 FSM_MOVE_STATE(INIT_HAL);
             } else { // EType::Wired
-                FSM_MOVE_STATE(MASTER_DISCOVERY);
+                FSM_MOVE_STATE(READY_FOR_MASTER_DISCOVERY);
             }
 
             skip_select = true;
         }
         break;
     }
+    case EState::READY_FOR_MASTER_DISCOVERY: {
+        auto certification_mode = bpl::cfg_get_certification_mode();
+        if (certification_mode < 0) {
+            LOG(ERROR) << "Failed reading certification_mode: " << certification_mode;
+            return false;
+        }
+        // In certification mode we want to wait till dev_set_config is received (wired backhaul)
+        // or start_wps_registration (wireless backhaul).
+        // Otherwise switch to MASTER_DISCOVERY
+        if (!certification_mode) {
+            FSM_MOVE_STATE(MASTER_DISCOVERY);
+        }
+
+        if (!m_agent_ucc_listener) {
+            LOG(ERROR) << "m_agent_ucc_listener == nullptr";
+            return false;
+        }
+
+        if (m_agent_ucc_listener->has_received_dev_set_config()) {
+            FSM_MOVE_STATE(MASTER_DISCOVERY);
+        }
+        break;
+    }
```

Comment:
Well, there is no really good reason to create a whole state for this.
You could simply do the whole thing on the "MASTER_DISCOVERY" state, and if the requirements are not met, just break, and the next time that the state will run it will try again.

```cpp
    case EState::MASTER_DISCOVERY: {
        ### After PR #1590 will be merged please take this flag from the AgentDB
        auto certification_mode = bpl::cfg_get_certification_mode();
        if (certification_mode < 0) {
            LOG(ERROR) << "Failed reading certification_mode: " << certification_mode;
            return false;
        }

        // In certification mode we want to wait till dev_set_config is received (wired backhaul)
        // or start_wps_registration (wireless backhaul).
        if (certification_mode && m_agent_ucc_listener && !m_agent_ucc_listener->has_received_dev_set_config()) {
            break;
        }

        // Rest of MASTER_DISCOVERY state...
    }
```

Reply:
Unfortunately it is not enough.
Even in certification mode dev_set_config is not always required.
We don't need it if WPS PBC is used or if the agent is run on the same machine where controller is run (we discussed `if (db->device_conf.local_controller && db->device_conf.local_gw) {` shortcut in the slack).

We still need an additional piece of data to check whether waiting for dev_set_config is required, and IMO an explicit state is better than a variable that acts like a state.

Reply:
You can add whatever validators you think we need. But you should remove the new state. 
A state of the backhaul manager should reflect a real logical state and not some specific scenario that is relevant only for certification.
If you want you can wrap it in a function and according to the function return value, decide if you continue with the state or break. That is also an option.
Either way, the state should be removed.

Reply:
I share @vitalii-komisarenko's opinion on this matter: despite required functionality can be achieved with some state variables, it deserves a new state (and the decision should not be taken based on if this new state is reached only in certification).